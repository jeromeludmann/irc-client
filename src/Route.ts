import { Action } from "redux";

export interface Route {
  serverKey: string;
  channelKey: string;
}

export interface RoutedAction extends Action<string> {
  route: Route;
}

export const STATUS = "@status";
export const RAW = "@raw";

export const BROADCAST_ALL = "@all";
export const BROADCAST_ACTIVE = "@active";
export const BROADCAST_NONE = "@none";

export const isRaw = (key: string) => key === RAW;
export const isStatus = (key: string) => key === STATUS;
export const isChannel = (key: string) => /^(&|#|\+|!)/.test(key);
export const isPrivate = (key: string) =>
  !isRaw(key) && !isStatus(key) && !isChannel(key);
