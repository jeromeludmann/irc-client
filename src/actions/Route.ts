import { Action } from "redux";

export interface Route {
  server: string;
  channel: string;
}

export type RoutedAction = Action & {
  route: Route;
};
