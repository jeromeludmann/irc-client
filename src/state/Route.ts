import { Action } from "redux";

export interface Route {
  server: string;
  channel: string;
}

export interface Routable extends Action<string> {
  route: Route;
}
