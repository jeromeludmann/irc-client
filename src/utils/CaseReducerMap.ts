export interface CaseReducerMap<R> {
  [action: string]: R
}
