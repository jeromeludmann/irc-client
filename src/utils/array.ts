export function push<T>(times: T[], time: T): Promise<number> {
  return new Promise(resolve => resolve(times.push(time)))
}

export function shift<T>(times: T[]): Promise<T> {
  return new Promise(resolve => resolve(times.shift()))
}
