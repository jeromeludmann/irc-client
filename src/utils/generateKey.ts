export function generateKey(existingServerKeys: string[]): string {
  const key = Math.random()
    .toString(36)
    .slice(2)
  return existingServerKeys.includes(key)
    ? generateKey(existingServerKeys)
    : key
}
