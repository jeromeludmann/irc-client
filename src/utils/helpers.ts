export const CRLF = '\r\n'

export const IRC_MESSAGE_LENGTH = 512 - CRLF.length

export const CHANNEL_REGEXP = /^([&|#|\+|!]\S+)$/

export function generateServerKey(existingServerKeys: string[]): string {
  const key = Math.random()
    .toString(36)
    .slice(2)
  return existingServerKeys.includes(key)
    ? generateServerKey(existingServerKeys)
    : key
}
