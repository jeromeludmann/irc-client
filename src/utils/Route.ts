import { Action } from 'redux'
import { CHANNEL_REGEXP } from '@app/utils/helpers'

export interface Route {
  serverKey: string
  bufferKey: string
}

export interface RoutedAction<T = string> extends Action<T> {
  route: Route
}

export const BufferKey = {
  STATUS: '@status',
  RAW: '@raw',
  ALL: '@all',
  ACTIVE: '@active',
  NONE: '@none',
}
export const isRaw = (key: string) => key === BufferKey.RAW

export const isStatus = (key: string) => key === BufferKey.STATUS

export const isChannel = (key: string) => CHANNEL_REGEXP.test(key)

export const isPrivate = (key: string) =>
  !isRaw(key) && !isStatus(key) && !isChannel(key)
