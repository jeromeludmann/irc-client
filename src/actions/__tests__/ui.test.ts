import {
  closeWindow,
  enterInputValue,
  goBackInputHistory,
  goForwardInputHistory,
  switchWindow,
  updateInputValue,
} from '@app/actions/ui'
import { Route } from '@app/utils/Route'

describe('ui actions', () => {
  const route: Route = { serverKey: 'serverKey', bufferKey: '#channel' }

  it('should create CloseWindowAction', () => {
    expect(closeWindow(route)).toMatchSnapshot()
  })

  it('should create EnterInputValueAction', () => {
    expect(enterInputValue('hello')).toMatchSnapshot()
  })

  it('should create GoBackInputHistoryAction', () => {
    expect(goBackInputHistory()).toMatchSnapshot()
  })

  it('should create GoForwardInputHistoryAction', () => {
    expect(goForwardInputHistory()).toMatchSnapshot()
  })

  it('should create SwitchWindowAction', () => {
    expect(switchWindow(route)).toMatchSnapshot()
  })

  it('should create UpdateInputValueAction', () => {
    expect(updateInputValue('hello')).toMatchSnapshot()
  })
})
