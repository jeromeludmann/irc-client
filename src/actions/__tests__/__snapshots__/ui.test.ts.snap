// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui actions should create CloseWindowAction 1`] = `
Object {
  "route": Object {
    "bufferKey": "#channel",
    "serverKey": "serverKey",
  },
  "type": "UI/CLOSE_WINDOW",
}
`;

exports[`ui actions should create EnterInputValueAction 1`] = `
Object {
  "payload": Object {
    "value": "hello",
  },
  "type": "UI/ENTER_INPUT_VALUE",
}
`;

exports[`ui actions should create GoBackInputHistoryAction 1`] = `
Object {
  "type": "UI/GO_BACK_INPUT_HISTORY",
}
`;

exports[`ui actions should create GoForwardInputHistoryAction 1`] = `
Object {
  "type": "UI/GO_FORWARD_INPUT_HISTORY",
}
`;

exports[`ui actions should create SwitchWindowAction 1`] = `
Object {
  "route": Object {
    "bufferKey": "#channel",
    "serverKey": "serverKey",
  },
  "type": "UI/SWITCH_WINDOW",
}
`;

exports[`ui actions should create UpdateInputValueAction 1`] = `
Object {
  "payload": Object {
    "value": "hello",
  },
  "type": "UI/UPDATE_INPUT_VALUE",
}
`;
